src/AWS.Deploy.Orchestration/CustomRecipeLocator.cs

```diff
@@ -0,0 +1,176 @@
+            _directoryManager = directoryManager;
+        }
+
+        /// <summary>
+        /// Wrapper method to fetch fetch custom recipe paths from a deployment-manifest file as well as
+        /// other locations that are monitored by the same source control root as the target application that needs to be deployed.
+        /// </summary>
+        /// <param name="targetApplicationFullPath">The absolute path to the csproj or fsproj file of the target application</param>
+        /// <returns>A <see cref="HashSet{String}"/> containing absolute paths of directories inside which the custom recipe snapshot is stored</returns>
+        public async Task<HashSet<string>> LocateCustomRecipePaths(string targetApplicationFullPath)
+        {
+            await LocateRecipePathsFromManifestFile(targetApplicationFullPath);
+
+            await LocateAlternateRecipePaths(targetApplicationFullPath);
+            
+            return new HashSet<string>(_customRecipePaths);
+        }
+
+        /// <summary>
+        /// Fetches custom recipe paths by parsing the deployment-manifest file that is associated with the target application.
+        /// </summary>
+        /// <param name="targetApplicationFullPath">The absolute path to the target application csproj or fsproj file</param>
+        private async Task LocateRecipePathsFromManifestFile(string targetApplicationFullPath)
+        {
+            List<string>? manifestFileRecipePaths;
+            try
+            {
+                manifestFileRecipePaths = await _deploymentManifestEngine.GetCustomRecipePaths(targetApplicationFullPath);
+            }
+            catch
+            {
+                _orchestratorInteractiveService.LogMessageLine(Environment.NewLine);
+                _orchestratorInteractiveService.LogErrorMessageLine("Failed to load custom deployment recommendations " +
+                   "from the deployment-manifest file due to an error while trying to deserialze the file.");
+                return;
+            }
+
+            foreach (var recipePath in manifestFileRecipePaths)
+            {
+                _customRecipePaths.Add(recipePath);
+            }
+        }
+
+        /// <summary>
+        /// Fetches custom recipe paths from other locations that are monitored by the same source control root
+        /// as the target application that needs to be deployed.
+        /// If the target application is not under source control then it scans the sub-directories of the solution folder for custom recipes.
+        /// </summary>
+        /// <param name="targetApplicationFullPath">The absolute path to the target application csproj or fsproj file</param>
+        private async Task LocateAlternateRecipePaths(string targetApplicationFullPath)
+        {
+            var targetApplicationDirectoryPath = _directoryManager.GetDirectoryInfo(targetApplicationFullPath).Parent.FullName;
+
+            if (await IsDirectoryUnderSourceControl(targetApplicationDirectoryPath))
```

Comment:
This code can be simplified. You should methods to determine the root directory based on source control or sln and then a single function that searches for recipes given that root directory. 

I don't like there is a complete separate implementation for sln and source control when the logic is the same for search after the root is determined.

Reply:
Fixed